File: exercises/js210/small_problems/medium1.js
// Rotation Part 1
function rotateArray(array) {
  if (!Array.isArray(array)) return undefined;

  array = array.slice()
  if (array.length > 0) array.push(array.shift());

  return array;
}

rotateArray([7, 3, 5, 2, 9, 1]);       // [3, 5, 2, 9, 1, 7]
rotateArray(['a', 'b', 'c']);          // ["b", "c", "a"]
rotateArray(['a']);                    // ["a"]
rotateArray([1, 'a', 3, 'c']);         // ["a", 3, "c", 1]
rotateArray([{ a: 2 }, [1, 2], 3]);    // [[1, 2], 3, { a: 2 }]
rotateArray([]);                       // []

// return `undefined` if the argument is not an array
rotateArray();                         // undefined
rotateArray(1);                        // undefined


// the input array is not mutated
const array = [1, 2, 3, 4];
rotateArray(array);                    // [2, 3, 4, 1]
array;                                 // [1, 2, 3, 4]


// Rotation Part 2
function rotateRightmostDigits(number, degrees) {
  let digits = String(number).split('');

  digits = digits.slice(0, digits.length - degrees)
                 .concat(rotateArray(digits.slice(digits.length - degrees),
                                                  digits.length));

  return parseInt(digits.join(''), 10);
}

rotateRightmostDigits(735291, 1);      // 735291
rotateRightmostDigits(735291, 2);      // 735219
rotateRightmostDigits(735291, 3);      // 735912
rotateRightmostDigits(735291, 4);      // 732915
rotateRightmostDigits(735291, 5);      // 752913
rotateRightmostDigits(735291, 6);      // 352917


// Rotation Part 3
function maxRotation(number) {
  let digits = String(number).split('');

  for (let i = 0; i < String(number).length; i += 1) {
    digits = digits.slice(0, i).concat(rotateArray(digits.slice(i, digits.length)));
  }

  return parseInt(digits.join(''), 10);
}

maxRotation(735291);          // 321579
maxRotation(3);               // 3
maxRotation(35);              // 53
maxRotation(105);             // 15 -- the leading zero gets dropped
maxRotation(8703529146);      // 7321609845


// Stack Machine Interpretation
function minilang(program) {
  let stack    = [];
  let register = 0;
  program = program.split(' ').map(command => (parseInt(command, 10) || command));

  program.forEach(command => {
    switch (command) {
      case 'PUSH':
        stack.push(register);
        break;
      case 'ADD':
        register += stack.pop();
        break;
      case 'SUB':
        register -= stack.pop();
        break;
      case 'MULT':
        register *= stack.pop();
        break;
      case 'DIV':
        register = Math.floor(register / stack.pop());
        break;
      case 'REMAINDER':
        register = Math.floor(register % stack.pop());
        break;
      case 'POP':
        register = stack.pop();
        break;
      case 'PRINT':
        console.log(register);
        break;
      default:
        register = command;
    }
  })
}

minilang('PRINT');
// 0

minilang('5 PUSH 3 MULT PRINT');
// 15

minilang('5 PRINT PUSH 3 PRINT ADD PRINT');
// 5
// 3
// 8

minilang('5 PUSH POP PRINT');
// 5

minilang('3 PUSH 4 PUSH 5 PUSH PRINT ADD PRINT POP PRINT ADD PRINT');
// 5
// 10
// 4
// 7

minilang('3 PUSH PUSH 7 DIV MULT PRINT');
// 6

minilang('4 PUSH PUSH 7 REMAINDER MULT PRINT');
// 12

minilang('-3 PUSH 5 SUB PRINT');
// 8

minilang('6 PUSH');
// (nothing is printed because the `program` argument has no `PRINT` commands)


// Word to Digit
const NUMBER_WORDS = { 'zero': 0, 'one': 1, 'two': 2, 'three': 3, 'four': 4,
                       'five': 5, 'six': 6, 'seven': 7, 'eight': 8, 'nine': 9 }

function wordToDigit(string) {
  Object.keys(NUMBER_WORDS).forEach((key) => {
    string.replace(new RegExp('\\b' + key + '\\b', 'g'), NUMBER_WORDS[key]);
  })

  return string;
}

wordToDigit('Please call me at five five five one two three four. Thanks.');


// Fibonacci Number Location By Length
function findFibonacciIndexByLength(targetLength) {
  let lastNums = [0n, 1n];
  let currentNum = 0n;
  let index = 1;

  while (String(currentNum).length < targetLength) {
    index += 1;
    currentNum = lastNums.reduce((a, b) => a + b);
    lastNums = [lastNums[1], currentNum];
  }

  console.log(index);
}

findFibonacciIndexByLength(2n) === 7n;    // 1 1 2 3 5 8 13
findFibonacciIndexByLength(3n) === 12n;   // 1 1 2 3 5 8 13 21 34 55 89 144
findFibonacciIndexByLength(10n) === 45n;
findFibonacciIndexByLength(16n) === 74n;
findFibonacciIndexByLength(100n) === 476n;
findFibonacciIndexByLength(1000n) === 4782n;
// findFibonacciIndexByLength(10000n) === 47847n;


// Fibonacci Numbers (Recursion)
function fibonacci(n) {
  if (n <= 1) return n;
  else        return fibonacci(n - 2) + fibonacci(n - 1);
}

fibonacci(1);       // 1
fibonacci(2);       // 1
fibonacci(3);       // 2
fibonacci(4);       // 3
fibonacci(5);       // 5
fibonacci(12);      // 144
fibonacci(20);      // 6765


// Fibonacci Numbers (Procedural)
function fibonacci(n) {
  let lastNums = [0n, 1n];
  let currentNum = 0n;

  for (let i = 1; i < n; i += 1) {
    currentNum = lastNums.reduce((a, b) => a + b);
    lastNums = [lastNums[1], currentNum];
  }

  return currentNum;
}

fibonacci(20);       // 6765
fibonacci(50);       // 12586269025
fibonacci(75);       // 2111485077978050


// Fibonacci Numbers (Memoization)
let fibNums = { 0: 0, 1: 1 };

let fibonacci2 = function fibonacci(n) {
  if (fibNums[n] !== undefined) {
    return fibNums[n];
  } else {
    return fibNums[n] = fibonacci2(n - 2) + fibonacci2(n - 1);
  }
}

console.log(fibonacci2(1));       // 1
console.log(fibonacci2(2));       // 1
console.log(fibonacci2(3));       // 2
console.log(fibonacci2(4));       // 3
console.log(fibonacci2(5));       // 5
console.log(fibonacci2(12));      // 144
console.log(fibonacci2(20));      // 6765

File: exercises/rb101-119/medium1.rb
#	Rotation (Part 1)
def rotate_array(arr)
  arr[1..-1] + [arr[0]]
end

def rotate_string(str)
  rotate_array(str.chars).join
end

def rotate_integer(int)
  rotate_array(int.digits.reverse).join
end

#	Rotation (Part 2)
def rotate_rightmost_digits(num, count)
  digits = num.to_s.chars.reverse
  (digits[0..-count - 1] + [rotate_array(digits[-count..-1])]).join
end

#	Rotation (Part 3)
def max_rotation(int)
  output = int.to_s
  int.digits.count.times do |i|
    output = rotate_rightmost_digits(output, int.digits.count - i)
  end
  output
end

# The rotation methods initially handled integers, as the original problem
# called for, but to satisfy the Further Exploration request to preserve zeroes,
# they were changed to output strings, as that is the only solution I can think
# of. I could then call to_i at the end, but that will still eliminate some
# number of zeroes--ie, 10001 would return 1010 instead of 01010.


#	1000 Lights
def switches(int)
  output = (1..int).to_a
  
  (2..int).each do |i|
    (2..int).each do |j|
      if j % i == 0
        output.include?(j) ? output.delete(j) : output << j
      end
    end
  end
  
  puts "Lights #{output[0..-2].join(", ")} and #{output[-1]} are on; "\
       "lights #{((1..int).to_a - output)[0..-2].join(", ")} and "\
       "#{((1..int).to_a - output)[-1]} are off."
end

#	Diamonds!
def diamond(int)
  spaces = (int - 1) / 2
  
  int.times do |i|
    puts (' ' * spaces.abs) + ('*' * (int - spaces.abs * 2)) + (' ' * spaces.abs)
    spaces -= 1
  end
end

def further_print_row(grid_size, distance_from_center)
  number_of_stars = grid_size - 2 * distance_from_center
  stars = ' ' * number_of_stars
  stars[0], stars[-1] = '*', '*'
  puts stars.center(grid_size)
end

def further_diamond(grid_size)
  max_distance = (grid_size - 1) / 2
  max_distance.downto(0) { |distance| further_print_row(grid_size, distance) }
  1.upto(max_distance)   { |distance| further_print_row(grid_size, distance) }
end

#	Stack Machine Interpretation
def minilang(input)
  register = 0
  stack = []
  input = input.downcase.split
  
  input.each do |command|
    case command
    when command.to_i.to_s then register = command.to_i
    when "push"            then stack << register
    when "add"             then register += stack.pop
    when "sub"             then register -= stack.pop
    when "mult"            then register *= stack.pop
    when "div"             then register /= stack.pop
    when "mod"             then register %= stack.pop
    when "pop"             then register = stack.pop
    when "print"           then puts register
    end
  end
end

def further_minilang(input)
  if input.count('^0-9()*/+-% ') > 0
    return "Invalid input. Only integers and operators are valid inputs."
  elsif input.length == 0
    return "Invalid input. Cannot process an empty stack."
  end
  processed_input = input.insert(0, '(') + ')'
  parentheses = processed_input
  while processed_input.include? ('(')
    parentheses = processed_input[processed_input[0..processed_input.index(")")]
                  .rindex("(")..processed_input.index(")")]

    processed_input[processed_input.index(parentheses)..processed_input
    .index(parentheses) + parentheses.length - 1] = pemdas(parentheses)
  end
  
  p processed_input
  return
end

def pemdas(equation)
  equation = equation.delete('()').split

  while equation.include?('**')
    i = equation.index('**')
    operation = equation[i]
    equation[i - 1..i + 1] =
    (equation[i - 1].to_i ** equation[i + 1].to_i).to_s
  end

  while equation.include?('*') || equation.include?('/') ||
        equation.include?('%')
    i = equation.index { |char| ['*', '/', '%'].include?(char) }
    operation = equation[i]
    case operation
    when '*'
      equation[i - 1..i + 1] =
      (equation[i - 1].to_i * equation[i + 1].to_i).to_s
    when '/'
      equation[i - 1..i + 1] =
      (equation[i - 1].to_i / equation[i + 1].to_i).to_s
    when '%'
      equation[i - 1..i + 1] =
      (equation[i - 1].to_i % equation[i + 1].to_i).to_s
    end
  end

  while equation.include?('+') || equation.include?('-')
    i = equation.index { |char| ['+', '-'].include?(char) }
    operation = equation[i]
    case operation
    when '+' 
      equation[i - 1..i + 1] =
      (equation[i - 1].to_i + equation[i + 1].to_i).to_s
    when '-'
      equation[i - 1..i + 1] =
      (equation[i - 1].to_i - equation[i + 1].to_i).to_s
    end
  end
  
  equation.join
end

#	Word to Digit
NUMBERS = { 'zero' => 0, 'one' => 1, 'two' => 2, 'three' => 3, 'four' => 4,
            'five' => 5, 'six' => 6, 'seven' => 7, 'eight' => 8, 'nine' => 9 }

def format_phone_number(str)
  if str.delete('^0-9').length == 10
    str = "(#{str[0..2]}) #{str[3..5]}-#{str[6..-1]}"
  end
  str
end

def word_to_digit(str)
  output       = str.split
  range_start  = output.index { |i| NUMBERS.keys.include?(i.downcase) }
  range_end    = range_start
  
  while range_start
    while output[range_end] != output.last &&
          NUMBERS.include?(output[range_end + 1].delete('^a-zA-Z'))
      range_end += 1
    end
    
    NUMBERS.keys.each do |i|
      output[range_start..range_end].each do |j|
        j.downcase! if NUMBERS[j.downcase]
        j.gsub!(/\b#{i}\b/, NUMBERS[i].to_s)
      end
    end
    
    output[range_start..range_end] = output[range_start..range_end].join
    output[range_start] = format_phone_number(output[range_start])
    
    range_start  = output.index { |i| NUMBERS.keys.include?(i) }
    range_end    = range_start
  end
  output.join(' ')
end

#	Fibonacci Numbers (Recursion)
def fibonacci(int, last_two = [1, 1])
  return last_two[1] if int <= 2
  last_two[0], last_two[1], = last_two[1], last_two[0] + last_two[1]
  fibonacci(int - 1, last_two)
end

#	Fibonacci Numbers (Procedural)
def fibonacci2(int)
  last_two = [1, 1]
  while int > 2
    last_two[0], last_two[1] = last_two[1], last_two[0] + last_two[1] 
    int -= 1
  end
  last_two[1]
end

#	Fibonacci Numbers (Last Digit)
def fibonacci_last(int)
  fibonacci2(int % 60).digits[0]
end